contrib/pipelinedb_kafka/pipelinedb_kafka.c

```diff
@@ -0,0 +1,895 @@
+	StartTransactionCommand();
+	load_consumer_state(id, &consumer);
+	copy = get_copy_statement(&consumer);
+	CommitTransactionCommand();
+
+	topic_conf = rd_kafka_topic_conf_new();
+	kafka = rd_kafka_new(RD_KAFKA_CONSUMER, NULL, err_msg, sizeof(err_msg));
+	rd_kafka_set_logger(kafka, logger);
+
+	/*
+	 * Add all brokers currently in pipeline_kafka_brokers
+	 */
+	if (consumer.brokers == NIL)
+		elog(ERROR, "no valid brokers were found");
+
+	foreach(lc, consumer.brokers)
+		valid_brokers += rd_kafka_brokers_add(kafka, lfirst(lc));
+
+	if (!valid_brokers)
+		elog(ERROR, "no valid brokers were found");
+
+	/*
+	 * Set up our topic to read from
+	 */
+	topic = rd_kafka_topic_new(kafka, consumer.topic, topic_conf);
+	if (rd_kafka_consume_start(topic, consumer.partition, consumer.offset) == -1)
+		elog(ERROR, "failed to start consuming: %s", rd_kafka_err2str(rd_kafka_errno2err(errno)));
+
+	messages = palloc0(sizeof(rd_kafka_message_t) * consumer.batch_size);
+
+	/*
+	 * Consume messages until we are terminated
+	 */
+	while (!got_sigterm)
+	{
+		ssize_t num_consumed;
+		int i;
+		int messages_buffered = 0;
+		StringInfoData buf;
+		bool xact = false;
+
+		num_consumed = rd_kafka_consume_batch(topic, consumer.partition,
+				CONSUMER_TIMEOUT, messages, consumer.batch_size);
+
+		if (num_consumed <= 0)
+			continue;
+
+		StartTransactionCommand();
+		xact = true;
+
+		initStringInfo(&buf);
+		for (i = 0; i < num_consumed; i++)
+		{
+			if (messages[i]->payload != NULL)
+			{
+				appendBinaryStringInfo(&buf, messages[i]->payload, messages[i]->len);
+				if (buf.data[buf.len] != '\n')
+					appendStringInfoChar(&buf, '\n');
+				messages_buffered++;
+			}
+			consumer.offset = messages[i]->offset;
+			rd_kafka_message_destroy(messages[i]);
+		}
+
+		/* we don't want to die in the event of any errors */
+		PG_TRY();
+		{
+			if (messages_buffered)
+				execute_copy(copy, &buf);
+		}
+		PG_CATCH();
+		{
+			elog(LOG, "[kafka consumer] %s <- %s failed to process message batch:", consumer.rel->relname, consumer.topic);
+			EmitErrorReport();
+			FlushErrorState();
+
+			AbortCurrentTransaction();
+			xact = false;
+			pg_usleep(1 * 1000 * 1000);
+		}
+		PG_END_TRY();
+
+		if (!xact)
+			StartTransactionCommand();
+
+		save_consumer_state(id, &consumer);
+
+		CommitTransactionCommand();
+	}
+
+  rd_kafka_topic_destroy(topic);
+  rd_kafka_destroy(kafka);
+  rd_kafka_wait_destroyed(CONSUMER_TIMEOUT);
+}
+
+/*
+ * create_consumer
+ *
+ * Create a row in pipeline_kafka_consumers representing a topic-relation consumer
+ */
+static Oid
+create_consumer(Relation consumers, text *relation, text *topic, text *format, text *delimiter, int batchsize)
+{
+	HeapTuple tup;
+	Datum values[CONSUMER_RELATION_NATTS];
+	bool nulls[CONSUMER_RELATION_NATTS];
+	Oid oid;
+
+	values[CONSUMER_ATTR_RELATION - 1] = PointerGetDatum(relation);
+	values[CONSUMER_ATTR_TOPIC - 1] = PointerGetDatum(topic);
+	values[CONSUMER_ATTR_PARTITION - 1] = Int32GetDatum(0);
```

Comment:
Yea that's the last thing I'm working on. The consumer proc will actually need to cycle through partitions after retrieving all of them from the Kafka cluster. I'll probably add some kind of `parallelism` arg to `kafka_consume_begin` so that each proc will read from `# partitions / # procs` partitions